src/main/java/com/pinterest/secor/io/impl/MessagePackSequenceFileReaderWriterFactory.java

```diff
@@ -161,5 +175,17 @@ public void write(KeyValue keyValue) throws IOException {
         public void close() throws IOException {
             this.mWriter.close();
         }
+
+        private int numberOfFieldsMappedInHeader(KeyValue keyValue) {
+            int fields = 1;
+
+            if (keyValue.hasKafkaKey())
+                fields++;
+
+            if (keyValue.hasTimestamp())
```

Comment:
I've added a test to check the scenario you are describing (a record with timestamp but no key) 
```
KeyValue kv4 =  (new KeyValue(23235, null, new byte[]{23, 45, 40 ,10, 122}, 1496318250));
```
This method is returning the number of fields that we are encoding together, not the field index. So for a record with offset but no key and no timestamp should return 1. If we had key, offset and timestamp it should return 3 and in case either the key or the timestamp is missing it will return 2.